File: statistical-analytics/scripts/llm_statistical_analysis.py
#!/usr/bin/env python3
"""Statistical analysis templates for LLM outputs.

This file assumes LLM outputs and evaluations already exist. It treats LLM
quality as measurable data: randomized prompt comparisons, factuality rates,
rubric scores, evaluator agreement, calibration, cost, latency, and drift.
"""

from __future__ import annotations


def optional_import(name: str):
    try:
        return __import__(name)
    except Exception:
        return None


pd = optional_import("pandas")
np = optional_import("numpy")


def require(module, package_name: str):
    if module is None:
        raise RuntimeError(f"Install {package_name} to run this recipe.")
    return module


def prompt_score_model(df):
    smf = require(optional_import("statsmodels.formula.api"), "statsmodels")
    return smf.ols("score ~ C(prompt_id) + input_tokens + output_tokens", data=df).fit()


def factuality_model(df):
    smf = require(optional_import("statsmodels.formula.api"), "statsmodels")
    return smf.logit("factual ~ C(prompt_id) + confidence + input_tokens + output_tokens", data=df).fit()


def brier_score(correct, probability):
    np_mod = require(np, "numpy")
    correct = np_mod.asarray(correct, dtype=float)
    probability = np_mod.asarray(probability, dtype=float)
    return np_mod.mean((correct - probability) ** 2)


def calibration_table(df, bins=5):
    pd_mod = require(pd, "pandas")
    out = df.copy()
    out["confidence_bin"] = pd_mod.qcut(out["confidence"], q=bins, duplicates="drop")
    return out.groupby("confidence_bin", observed=True).agg(
        mean_confidence=("confidence", "mean"),
        observed_accuracy=("factual", "mean"),
        n=("factual", "size"),
    )


def evaluator_agreement(labels_a, labels_b):
    sklearn_metrics = require(optional_import("sklearn.metrics"), "scikit-learn")
    return sklearn_metrics.cohen_kappa_score(labels_a, labels_b)


def cost_quality_frontier(df):
    return df.groupby(["prompt_id", "model_id"], as_index=False).agg(
        mean_score=("score", "mean"),
        factuality_rate=("factual", "mean"),
        mean_latency_ms=("latency_ms", "mean"),
        mean_cost_usd=("cost_usd", "mean"),
    )


def embedding_clusters(embedding_matrix, n_clusters=5):
    require(optional_import("sklearn"), "scikit-learn")
    from sklearn.cluster import KMeans
    from sklearn.preprocessing import StandardScaler

    X = StandardScaler().fit_transform(embedding_matrix)
    return KMeans(n_clusters=n_clusters, n_init="auto", random_state=123).fit(X)


def drift_table(df):
    out = df.copy()
    out["eval_date"] = pd.to_datetime(out["evaluated_at"]).dt.date
    return out.groupby(["eval_date", "prompt_id", "model_version"], as_index=False).agg(
        mean_score=("score", "mean"),
        factuality_rate=("factual", "mean"),
        mean_confidence=("confidence", "mean"),
        mean_latency_ms=("latency_ms", "mean"),
        mean_cost_usd=("cost_usd", "mean"),
    )


if __name__ == "__main__":
    print("Import this file after creating an LLM evaluation table with prompt_id, item_id, score, factual, and confidence.")

File: statistical-analytics/scripts/llm_statistical_analysis.R
#!/usr/bin/env Rscript

# Statistical analysis templates for LLM outputs.
# This file does not call an LLM API. It assumes outputs have already been
# generated and evaluated, then treats them as experimental/statistical data.

llm_schema <- data.frame(
  column = c(
    "item_id", "prompt_id", "model_id", "model_version", "output_id",
    "evaluator_id", "score", "factual", "confidence", "input_tokens",
    "output_tokens", "latency_ms", "cost_usd", "evaluated_at"
  ),
  meaning = c(
    "Stable test item or task identifier",
    "Prompt or system design assigned to the item",
    "Model family or provider label",
    "Model version string captured at evaluation time",
    "Unique output identifier",
    "Human or model evaluator identifier",
    "Rubric score, usually numeric or ordinal",
    "Binary factuality or correctness label",
    "Stated or evaluator-derived confidence from 0 to 1",
    "Input token count",
    "Output token count",
    "Observed latency in milliseconds",
    "Estimated cost in dollars",
    "Timestamp of evaluation"
  )
)

example_llm_data <- data.frame(
  item_id = rep(sprintf("item_%02d", 1:30), each = 2),
  prompt_id = rep(c("baseline", "candidate"), times = 30),
  model_id = "example_model",
  model_version = "v1",
  output_id = sprintf("out_%03d", 1:60),
  evaluator_id = "human_1",
  score = c(rep(c(3, 4), 15), rep(c(4, 5), 15)),
  factual = rep(c(1, 1, 0, 1, 1), length.out = 60),
  confidence = pmin(0.98, rep(c(0.62, 0.75, 0.58, 0.83, 0.91, 0.69), length.out = 60) + rep(seq_len(30) %% 5, each = 2) * 0.005),
  input_tokens = 380 + rep(seq_len(30), each = 2) + rep(c(0, 12), times = 30),
  output_tokens = 150 + rep(seq_len(30) %% 7, each = 2) + rep(c(0, 15), times = 30),
  latency_ms = seq(900, 1600, length.out = 60),
  cost_usd = seq(0.002, 0.006, length.out = 60),
  evaluated_at = Sys.time(),
  stringsAsFactors = FALSE
)

prompt_score_model <- function(data) {
  stats::lm(score ~ prompt_id + input_tokens + output_tokens, data = data)
}

factuality_model <- function(data) {
  stats::glm(factual ~ prompt_id + confidence + input_tokens + output_tokens, data = data, family = binomial())
}

prompt_ab_test <- function(data) {
  stats::prop.test(
    x = with(data, tapply(factual, prompt_id, sum)),
    n = table(data$prompt_id)
  )
}

brier_score <- function(correct, probability) {
  mean((correct - probability)^2, na.rm = TRUE)
}

calibration_table <- function(data, bins = 5) {
  breaks <- unique(quantile(data$confidence, probs = seq(0, 1, length.out = bins + 1), na.rm = TRUE))
  data$confidence_bin <- cut(data$confidence, breaks = breaks, include.lowest = TRUE)
  aggregate(
    cbind(mean_confidence = confidence, observed_accuracy = factual) ~ confidence_bin,
    data = data,
    FUN = mean
  )
}

cost_quality_frontier <- function(data) {
  aggregate(
    cbind(score, factual, latency_ms, cost_usd) ~ prompt_id + model_id,
    data = data,
    FUN = mean
  )
}

cohen_kappa_2_raters <- function(labels_a, labels_b) {
  tab <- table(labels_a, labels_b)
  observed <- sum(diag(tab)) / sum(tab)
  expected <- sum(rowSums(tab) * colSums(tab)) / sum(tab)^2
  (observed - expected) / (1 - expected)
}

embedding_cluster_template <- function(embedding_matrix, centers = 5) {
  stats::kmeans(scale(embedding_matrix), centers = centers, nstart = 20)
}

drift_table <- function(data) {
  data$eval_date <- as.Date(data$evaluated_at)
  aggregate(
    cbind(score, factual, confidence, latency_ms, cost_usd) ~ eval_date + prompt_id + model_version,
    data = data,
    FUN = mean
  )
}

if (identical(environment(), globalenv())) {
  print(llm_schema)
  print(summary(prompt_score_model(example_llm_data)))
  print(summary(factuality_model(example_llm_data)))
  print(prompt_ab_test(example_llm_data))
  print(calibration_table(example_llm_data))
  print(brier_score(example_llm_data$factual, example_llm_data$confidence))
  print(cost_quality_frontier(example_llm_data))
}
